### ISSUE TITLE
Unable to compile Erlang 22.0.7 on macOS 10.15

### ISSUE THREAD

**Issue by schlenks** [kerl/kerl#320]
An issue was posted on asdf-vm/asdf-erlang/issues/105 about erlang being unable to build on the upcoming version of macOS 10.15. Since asdf-vm uses kerl underneath, it was suggested an issue be created here.

Here is the output from trying to install:

```
Building Erlang/OTP 22.0.7 (asdf_22.0.7), please wait...
Configure options have changed. Reconfiguring...
APPLICATIONS DISABLED (See: /Users/schlenks/.asdf/plugins/erlang/kerl-home/builds/asdf_22.0.7/otp_build_22.0.7.log)
 * crypto         : No usable OpenSSL found
 * ssh            : No usable OpenSSL found
 * ssl            : No usable OpenSSL found

Build failed.
make[3]: *** [../ebin/hipe_tagscheme.beam] Segmentation fault: 11
make[3]: *** [../ebin/hipe_rtl_mk_switch.beam] Segmentation fault: 11
make[3]: *** [../ebin/hipe_rtl_cleanup_const.beam] Segmentation fault: 11
make[3]: *** [../ebin/hipe_rtl_binary.beam] Segmentation fault: 11
make[3]: *** [../ebin/hipe_rtl_binary_match.beam] Segmentation fault: 11
make[3]: *** [../ebin/hipe_icode2rtl.beam] Segmentation fault: 11
make[3]: *** [../ebin/hipe_rtl_liveness.beam] Segmentation fault: 11
make[2]: *** [opt] Error 2
make[1]: *** [opt] Error 2
make: *** [secondary_bootstrap_build] Error 2
```

I've looked at it a bit and tried a few different tactics to get it to recognize the openssl library:
1) installed openssl 1.0.2s via brew
2) installed openssl 1.1.1c via brew
3) compiled openssl 1.0.2.s from source code

`KERL_CONFIGURE_OPTIONS` set in `.zshrc`:
```
export KERL_CONFIGURE_OPTIONS="--disable-debug --disable-silent-rules --without-javac --enable-shared-zlib --enable-dynamic-ssl-lib --enable-sctp --enable-smp-support --enable-threads --enable-kernel-poll --enable-wx --enable-darwin-64bit --with-ssl=/usr/local/opt/openssl@1.0/bin
```

**Comment by mrallen1** [comment]
@sanmiguel This really could use some TLC if you have time/desire to pick it up.

**Comment by sanmiguel** [comment]
We have some magic in place on macOS to autodetect openssl from brew if
it's installed there.

IIRC that doesn't happen if you've got `--with-ssl=...` in
`$KERL_CONFIGURE_OPTS`

Try removing the `--with-ssl` opt and rerun the install with a
brew-installed openssl present.

**Comment by schlenks** [comment]
@sanmiguel 
my `KERL_CONFIGURE_OPTIONS`:
```
export KERL_CONFIGURE_OPTIONS="--enable-debug --disable-silent-rules --without-javac --enable-shared-zlib --enable-dynamic-ssl-lib --enable-hipe --enable-sctp --enable-smp-support --enable-threads --enable-kernel-poll --enable-wx --enable-darwin-64bit"
```
output from attempt:
```
Building Erlang/OTP 22.0.7 (asdf_22.0.7), please wait...
Configure options have changed. Reconfiguring...
APPLICATIONS DISABLED (See: /Users/schlenks/.asdf/plugins/erlang/kerl-home/builds/asdf_22.0.7/otp_build_22.0.7.log)
 * crypto         : No usable OpenSSL found
 * ssh            : No usable OpenSSL found
 * ssl            : No usable OpenSSL found

Build failed.
make[3]: *** [../ebin/hipe_rtl_mk_switch.beam] Segmentation fault: 11
make[3]: *** [../ebin/hipe_icode2rtl.beam] Segmentation fault: 11
make[3]: *** [../ebin/hipe_rtl_verify_gcsafe.beam] Segmentation fault: 11
make[3]: *** [../ebin/hipe_tagscheme.beam] Segmentation fault: 11
make[3]: *** [../ebin/hipe_rtl_binary_match.beam] Segmentation fault: 11
make[3]: *** [../ebin/hipe_rtl_cfg.beam] Segmentation fault: 11
make[3]: *** [../ebin/hipe_rtl_symbolic.beam] Segmentation fault: 11
make[2]: *** [opt] Error 2
make[1]: *** [opt] Error 2
make: *** [secondary_bootstrap_build] Error 2

```

**Comment by mrallen1** [comment]
I have this memory tickling my brain that we had some kind of similar problem a while ago - last year maybe and the root cause turned out to be that brew itself was not installed correctly or something along those lines.  Do you have any recollection of this @sanmiguel?

**Comment by sanmiguel** [comment]
Doesn't ring a bell, sorry.

One thing worth checking is whether asdf is actually using your kerlrc. I
set mine up a long time ago, and can't remember how it was by default but I
know that I have a symlink to ~/.kerlrc from inside the $ASDF_DIR
somewhere, so asdf uses it. I'm not at my laptop right now but I'll check
it tomorrow and try to write it up a bit.

**Comment by lsdiamonds** [comment]
this was the fix last time https://github.com/kerl/kerl/pull/301/commits/ea951d4dfd10e5b80f8a7a1bc7a56062fd426907#diff-75743bad0452412b658dabebd6b439f4

**Comment by mrallen1** [comment]
@schlenks Any desire to test PR #324 to see if it resolves your issue?

**Comment by pdgonzalez872** [comment]
@mrallen1 as you have pointed out in the readme, `brew` doesn't symlink to
`openssl`. I'm not 100% sure what changed or why this seems to have become an
issue recently again. I've seen this come up a couple of times at work and some
folks are discussing it in issues in different projects.

The solution that worked for us was to explicitly set the `--with-ssl` with a
path, in our case: `--with-ssl=/usr/local/Cellar/openssl/1.0.2p`.

There is also some indirection here: I think some other common tool that uses
`openssl` and symlinks it correctly. But, in a vanilla install these days with
minimal `brew` usage, things will remain unlinked. The ones that have the
symlink will work correctly, the ones that don't will eventually run into the
`SSL` error and after some reseach will pass in the `--with-ssl=<path>` option.

I believe the root cause is that the default we use for `openssl` is not the correct path and wh

A couple of solutions come to mind:

1) append `/bin` or `/include` to this line:
https://github.com/kerl/kerl/blob/master/kerl#L644 like `brew` suggests
https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/link.rb#L104-L113.
I'm unsure if this would do it, but am glad to help. I'd most likely need some
hand holding when testing things correctly (using the right versions of
things). This would be awesome. :)

2) add to https://github.com/kerl/kerl#compiling-crypto-on-macs. We have a note
already, but I saw some folks using the incorrect path in an issue somewhere.

3) add a note to a big source of the issues `asdf-erlang` about this SSL issue
and solution (adding the `--with-ssl`) flag

For 2) and 3) I think the following could do it. If you would like, I'll create
a patch for this repo with this info:

> You may encounter an SSL error with an output along these lines:
>
> * crypto         : No usable OpenSSL found
> * ssh            : No usable OpenSSL found
> * ssl            : No usable OpenSSL found
>
> If you see this, you can use the `--with-ssl` flag with a path (in this case,
> the path is `1.0.2p`):
> `--with-ssl=/usr/local/Cellar/openssl/1.0.2p`

Thanks for maintaining this tool. Happy to help in any way I can.

**Comment by andykent** [comment]
I am also hitting issues using `asdf-erlang` on 10.15. Once you get past the SSL stuff
https://github.com/asdf-vm/asdf-erlang/issues/116

I think the recent 1.8.6 changes addresses this and it looks like `kerl 1.8.6` is released? but it wasn't tagged in the git so the URL `https://raw.githubusercontent.com/kerl/kerl/1.8.6/kerl` doesn't work. I think this would need fixing before asdf-erlang` could be easily updated.

**Comment by Stratus3D** [comment]
Yes tagging 1.8.6 would be great. If there is not a tag for 1.8.6 I guess I could hardcode the commit hash for 1.8.6 in asdf-erlang, but I'd rather not have to do that.

**Comment by mrallen1** [comment]
I forgot to tag it. That's been fixed now.

**Comment by Stratus3D** [comment]
Awesome! Thanks so much @mrallen1! kerl is a life saver!

**Comment by andykent** [comment]
Thanks so much for the quick fix @mrallen1 and for all your work on kerl. 🙌

**Comment by andykent** [comment]
@mrallen1 Just reporting back here, I think there are still issues with building on 10.15. More details here https://github.com/asdf-vm/asdf-erlang/issues/116#issuecomment-538679935

I'm more than happy to help try and get this resolved. I'm not a compilation expert but please let me know if there is any way I can help out with further debugging.

**Comment by dc0d** [comment]
Trying to `asdf install erlang 22.1.3` on MacOS Catalina. Set `export PATH="/usr/local/Cellar/openssl/1.0.2t/bin:$PATH"` and `export KERL_CONFIGURE_OPTIONS="--without-javac --with-ssl=/usr/local/Cellar/openssl/1.0.2t"`. Build failed.

Still getting:

```
 * crypto         : No usable OpenSSL found
 * jinterface     : No Java compiler found
 * ssh            : No usable OpenSSL found
 * ssl            : No usable OpenSSL found
```

Am I missing something?

**Comment by mrleeio** [comment]
I've been following the maze of links throughout these Catalina issues and what I really need is a rough idea of how / when the compilation issues might be addressed.

**Background:**

My workplace uses Erlang v19.3 and none of the workarounds I've found work when trying to install on 10.15.

Things I have tried:

- Changing build target to OSX 10.14
- Modifying KERL_CONFIG_OPTIONS

All of this reading leads me to believe this is a bug present in Command Line tools.  Is Kerl working on a solution?  Are we collectively waiting on Apple for a patch?  Are older versions of Erlang going to see the necessary updates to build on 10.15?

Is there something I, as an individual contributor, could do to help?

**Comment by mrallen1** [comment]
Hi. In Catalina, apparently Apple enabled "stack checking" in its clang toolchain which makes erlc segfault. The solution seems to be `export CFLAGS="-fno-stack-check $CFLAGS"` which *has* been added to Erlang 22.1.4 but it is unlikely to be backported to older Erlang releases like 19.x. 

Kerl **will** include this (or some other workaround) at some point but none of us have a Catalina machine to do local testing (yet)

If you'd like to help us, I will update my PR #326 with this solution and you can test it?

**Comment by mrleeio** [comment]
> If you'd like to help us, I will update my PR #326 with this solution and you can test it?

Absolutely.

Note this is the error I experience when using the above export, which I forgot to mention in my first post.

```
~ % kerl build 19.3
Verifying archive checksum...
Checksum verified (a8c259ec47bf84e77510673e1b76b6db)
Extracting source code
Building Erlang/OTP 19.3 (19.3), please wait...
DOCUMENTATION INFORMATION (See: /Users/mrleeio/.kerl/builds/19.3/otp_build_19.3.log)
 * documentation  :
 *                  fop is missing.
 *                  Using fakefop to generate placeholder PDF files.

Build failed.
Makefile:72: warning: overriding commands for target `clean'
/Users/mrleeio/.kerl/builds/19.3/otp_src_19.3/make/otp_subdir.mk:29: warning: ignoring old commands for target `clean'
=== Entering application hipe
 VSN	hipe.hrl
 ERLC	../ebin/hipe_rtl.beam
erts_mmap: Failed to create super carrier of size 512 MB
make[3]: *** [../ebin/hipe_rtl.beam] Error 1
make[2]: *** [opt] Error 2
make[1]: *** [opt] Error 2
make: *** [secondary_bootstrap_build] Error 2

Please see /Users/mrleeio/.kerl/builds/19.3/otp_build_19.3.log for full details.
```

**Comment by mrleeio** [comment]
For anyone curious about Erlang > 21 I'm cross posting this from a pull-request:

> Found a Japanese blog entry which explores the issue in more depth.
>
> https://qiita.com/irisTa56/items/21aa6dfe2402c6ba6abc
> 
> TL;DR this pull-request will only work for Erlang > v21, anything older which requires hipe will continue to fail.
> 
> Potential work around is to use the following kerl config:
> 
> `KERL_CONFIGURE_OPTIONS="--disable-hipe --with-ssl=/usr/local/opt/openssl"`
>
> In combination with the stack-check disabled:
> 
> `export CFLAGS="-O2 -g -fno-stack-check"`
>
> Will try it out when I get home.

**Comment by tatsuya6502** [comment]
I ran into the same issue with ASDF with Erlang 22.1.4. And doing the followings solved my issue.

Install Homebrew `openssl@1.1` and other dependencies. 

```console
$ brew install autoconf automake libtool openssl@1.1 wxmac
```

(I took these package names form Homebrew's [erlang formula](https://github.com/Homebrew/homebrew-core/blob/cbfeb1307849d1a89d22d5556d1acacad2040fe7/Formula/erlang.rb#L16-L20))

Create `.kerlrc` in asdf-erlang's kerl home.

```console
$ mkdir -p ~/.asdf/plugins/erlang/kerl-home/

$ echo 'KERL_CONFIGURE_OPTIONS="--without-javac --with-ssl=/usr/local/opt/openssl@1.1"' &
      > ~/.asdf/plugins/erlang/kerl-home/.kerlrc
```

Install Erlang 22.1.4.

```console
$ asdf install erlang 22.1.4  
Extracting source code
Building Erlang/OTP 22.1.4 (asdf_22.1.4), please wait...
APPLICATIONS DISABLED (See: /Users/tatsuya/.asdf/plugins/erlang/kerl-home/builds/asdf_22.1.4/otp_build_22.1.4.log)
 * jinterface     : Java compiler disabled by user
 * odbc           : ODBC library - header check failed

DOCUMENTATION INFORMATION (See: /Users/tatsuya/.asdf/plugins/erlang/kerl-home/builds/asdf_22.1.4/otp_build_22.1.4.log)
 * documentation  : 
 *                  fop is missing.
 *                  Using fakefop to generate placeholder PDF files.

Erlang/OTP 22.1.4 (asdf_22.1.4) has been successfully built
Installing Erlang/OTP 22.1.4 (asdf_22.1.4) in /Users/tatsuya/.asdf/installs/erlang/22.1.4...
You can activate this installation running the following command:
. /Users/tatsuya/.asdf/installs/erlang/22.1.4/activate
Later on, you can leave the installation typing:
kerl_deactivate
Cleaning up compilation products for 
Cleaned up compilation products for  under /Users/tatsuya/.asdf/plugins/erlang/kerl-home/builds

Erlang 22.1.4 has been installed. Activate globally with:

    asdf global erlang 22.1.4

Activate locally in the current folder with:

    asdf local erlang 22.1.4


```

Ensure that Erlang works.

```console
$ asdf local erlang 22.1.4
$ erl
Erlang/OTP 22 [erts-10.5.3] [source] [64-bit] [smp:2:2] [ds:2:2:10] [async-threads:1] [hipe]

Eshell V10.5.3  (abort with ^G)
1> ^G
User switch command
 --> q
```

**Comment by miguelcoba** [comment]
This worked for me on Catalina 10.15.1

# MacOS Catalina (10.15.1)
# There is an issue in the compilation phase of the process. The workaround is this:
export KERL_CONFIGURE_OPTIONS="--without-javac --with-ssl=$(brew --prefix openssl)"
export CFLAGS="-O2 -g -fno-stack-check"
asdf install erlang 22.1
asdf glboal erlang 22.1

**Comment by jemmrich** [comment]
I just went through this myself on MacOS Catalina 10.15.2. I tried everything above but it wasn't working. One thing that seemed to help was removing and reinstalling the erlang plugin and any existing erlang installations. Here is the full process I went through:

```
brew reinstall coreutils automake autoconf openssl libyaml readline libxslt libtool unixodbc
brew link coreutils automake autoconf openssl libyaml readline libxslt libtool unixodbc

asdf plugin remove erlang     
asdf plugin add erlang
asdf uninstall erlang 22.1

export KERL_CONFIGURE_OPTIONS="--disable-hipe --with-ssl=$(brew --prefix openssl)"
export CFLAGS="-O2 -g -fno-stack-check"

asdf install erlang 22.1
```

**Comment by joevandyk** [comment]
@jemmrich that worked for, thanks!

**Comment by T0ha** [comment]
Hi
It looks like for older versions, like 19.x `--without-hipe` is needed instead of `--disable-hipe`, despite in `./configure --help` it is written to use the latter.

**Comment by paulo-ferraz-oliveira** [comment]
I was able to compile using MacOS 10.15 and OTP 22.0.7, as shown in this (adapted) CI run: https://github.com/paulo-ferraz-oliveira/kerl/actions/runs/3579523699/jobs/6020807708.

**Comment by jadeallenx** [comment]
going to close this now, reopen if still needed
